### PR TITLE
feature/WEB-3799: add JWT grant support

### PR DESCRIPTION
**Pull request by fredmajor**
Scope:
- add support for JWT Bearer token grant type (for communication with Salesforce)

- data model changes: use naming conventions and field lengths aligned with `oauth2_provider.Application` class. Client should support what Provider can produce.

- data model changes: add fields for JWT grant type, denormalize AccessToken from previous one-json-column approach to a structured format

- add token re-fetch feature, when token expiry detected. The feature includes a safety fuse - if the token is attempted for a re-fetch in <10s period, an exception is thrown, to avoid spamming the provider because of likely our fault

- completed TODOs from previous PR

@avkoval (as Security Officer) this PR contains an utility used to sign a claim we send to Salesforce to obtain an Oauth token. Please review [this file](https://github.com/Livit/Labster.OAuth2Client/blob/4dc39a7231eb3af5e3a12716ab7a0ef697731855/oauth2_client/utils/crypto.py) 

Follow up work:
- reiterate on SF OAuth setup: Please also add steps for DevOps to setup JWT case in a step-by-step manner that can be copy-pasted or linked to them in future Jira task to connect LS with SF.
- separate PR: add pip requirements to client's setup.py
- separate PR: remove edx_lint, bump pylint version
- ~~(optional): remove JSONFields to improve backward compatibility~~; [jira](https://liv-it.atlassian.net/browse/WEB-3919?atlOrigin=eyJpIjoiOTVmOTdmMDY5ZmJiNDRjYTkxMDQ4Mzg0MGY4MDVjMGMiLCJwIjoiaiJ9) 
- (optional, low priority) restructure tests, to enable use of test-only model classes; [jira](https://liv-it.atlassian.net/browse/WEB-3920?atlOrigin=eyJpIjoiZDQzZmRmODBmOGZjNDcyNWFlOTk4NjMyMTBiZWY3ZGMiLCJwIjoiaiJ9)

Full documentation: https://liv-it.atlassian.net/wiki/spaces/WEB/pages/891191387/OAuth2Client+documentation

Related Jiras:
- https://liv-it.atlassian.net/browse/WEB-3799?atlOrigin=eyJpIjoiYTRiYmZlNDcyODllNGY0Nzg1ZTA3NWJjOTg4YWYwMjYiLCJwIjoiaiJ9

### CODE REVIEW

**Comment by labstersteve** [comment]
Full marks for a lucid explanation of what the delta does. Beginning review ...

File: oauth2_client/client.py [hunk 1]
```diff
@@ -29,9 +38,58 @@ class OAuth2Client(OAuth2Session):
...
+        :param expired_callback: (optional) function accepting one parameter, `Application`. It is
+            called when TokenExpiredError is detected or 403 Forbidden status code received.
+            It should return a valid (not expired) oauth2_client.model.AccessToken object for
+            communication with Resource Owner specified by `Application`.
+            Explanation: The parent class, OAuth2Session already implements automated token refresh,
+            but their solution works only for flows that issue a refresh_token. Both currently
+            supported flows, JWT and Client Credentials don't provide the refresh_token, so we have
+            to handle the refresh ourselves. (Client Credentials MAY issue refresh_token but it
+            doesn't in case of the `oauth2_provider` we use). If the token expiry is detected more
+            than once within `REAUTH_LIMIT_SECONDS` period, `oauthlib.oauth2.TokenExpiredError`
```
Comment (labstersteve) [inline oauth2_client/client.py]: Is token expiry the only possible cause of the stated condition? I ask because it's important not to mislead future readers with poor naming. Also, have we correctly accounted for timeouts in the server connection? This is another condition where a circuit breaker can stop clients repeatedly beating on the server by returning immediate failure as long as the breaker remains tripped.

File: oauth2_client/client.py [hunk 2]
```diff
@@ -29,9 +38,58 @@ class OAuth2Client(OAuth2Session):
...
+            is raised (safety fuse approach).
+        """
+        self.app = token.application  # Application this client talks to
+        self.last_token_fetch_time = None  # used for the safety fuse preventing too often re-authorization
```
Comment (labstersteve) [inline oauth2_client/client.py]: Since the only thing the code fetches is tokens we could probably use `last_fetch_time`, with two fewer syllables to mentally pronounce as you read the code. Once they become habitual these small things can really add up, even though each alone is tiny. Or maybe as a compromise (and much better if more than one thing is being fetched) `last_token_time` ...

File: oauth2_client/client.py [hunk 3]
```diff
@@ -29,9 +38,58 @@ class OAuth2Client(OAuth2Session):
...
+        Wraps provided expiry callback into a safety fuse, so we won't spam the provider with token
+        requests in case of problems with the client.
+
+        :param expired_callback: a function taking one argument, `oauth2_client.models.Application`.
+            It has to fetch and store new AccessToken.
+        :return:  a wrapped callback function
+        """
+        def wrapped_callback(app):
+            if self.last_token_fetch_time \
+                    and self.last_token_fetch_time + timedelta(seconds=self.REAUTH_LIMIT_SECONDS) >= timezone.now():
```
Comment (labstersteve) [inline oauth2_client/client.py]: This ugliness copes with the single startup case where there has been no previous fetch.

Simpler rather to initialise `self.last_token_time` to now minus one minute,* and then the `None` test can be omitted, leading to a simpler logic flow and a truer assertion of initialization conditions: "No token requests in the last 60 seconds" is better than "No token requests yet." Explicit is better than implicit.

* In fact, `now() - REAUTH_LIMT_SECONDS`

File: oauth2_client/tests/test_flow_jwt.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+        app_data = {
+            'authorization_grant_type': 'jwt-bearer',
+            'client_secret': self.TEST_KEY,
+            'extra_settings': {'subject': 'xyz@abc.com.lightning'},
+            'token_uri': 'https://test.salesforce.com/services/oauth2/token',
+        }
+        app = ApplicationFactory(**app_data)
+        # mock provider's response
+        m.post(
```
Comment (auraz) [inline oauth2_client/tests/test_flow_jwt.py]: Please do not use 1 and 2 chars variables. Should be in pylint settings.

File: oauth2_client/tests/test_flow_client_credentials.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+from django.utils import timezone as tz
+
+from test_case import StandaloneAppTestCase
+
+
+class ClientCredentialsFlowTest(StandaloneAppTestCase):
+    """
+    Tests for Client Credentials Flow
+    """
+    CURR_DIR = os.path.dirname(os.path.abspath(__file__))
```
Comment (auraz) [inline oauth2_client/tests/test_flow_client_credentials.py]: Where it is needed?

File: oauth2_client/client.py [hunk 6]
```diff
@@ -43,63 +101,68 @@ def request(self, method, url, *args, **kwargs):  # pylint: disable=arguments-di
...
+        try:
+            # super().request may raise `TokenExpiredError`
+            resp = super(OAuth2Client, self).request(method, absolute_url, *args, **kwargs)
+            if resp.status_code == 403:
+                raise TokenExpiredError(description="403 Forbidden status code response received")
+            else:
+                return resp
+        except TokenExpiredError:
+            log.debug("Access Token for app=%s expired.", self.app.name)
+            if not self.expired_callback:
```
Comment (auraz) [inline oauth2_client/client.py]: Please point cases where we instantiate Clients with a not-none expired callback?

File: oauth2_client/client.py [hunk 7]
```diff
@@ -43,63 +101,68 @@ def request(self, method, url, *args, **kwargs):  # pylint: disable=arguments-di
...
         """
         absolute_url = urljoin(self.service_host, url)
-        return super(OAuth2Client, self).request(method, absolute_url, *args, **kwargs)
-
-
-def get_client(service_name):
+        try:
+            # super().request may raise `TokenExpiredError`
+            resp = super(OAuth2Client, self).request(method, absolute_url, *args, **kwargs)
+            if resp.status_code == 403:
```
Comment (auraz) [inline oauth2_client/client.py]: Doublechecking. Could 403 error to come due to another source other than `access token is expired`, or it is only raised when the token is expired?

Comment (auraz) [reply oauth2_client/client.py]: And if you have status code, why you need exception?

File: oauth2_client/client.py [hunk 8]
```diff
@@ -43,63 +101,68 @@ def request(self, method, url, *args, **kwargs):  # pylint: disable=arguments-di
...
 
-    Arguments:
-        access_token (object): oauth2_client.models.AccessToken instance.
-    """
-    return time.time() > access_token['expires_at'] - TIMEOUT_SECONDS
+    return OAuth2Client(token, expired_callback=fetch_and_store_token)
 
 
-def _get_token(app):
+def fetch_and_store_token(app):
```
Comment (auraz) [inline oauth2_client/client.py]: I do not need we need this as separate function, but maybe you need it in tests?

File: oauth2_client/tests/test_client.py [hunk 9]
```diff
@@ -1,84 +1,204 @@
...
+        fetch_token_mock.return_value = token_from_provider
 
         client = get_client(app.name)
         self.assertIsNotNone(client)
-        access_token = AccessToken.objects.get(application=app)
-        self.assertEqual(token, access_token.token)
-        get_token_mock.assert_called_once_with(app)
+        token_from_db = AccessToken.objects.get(application=app)
+        self.assertEqual(token_from_provider.token, token_from_db.token)
+        self.assertEqual(
```
Comment (auraz) [inline oauth2_client/tests/test_client.py]: Please fit into one line if it fits 120, it is much easier to read and review because that one line will have single semantic meaning.

**Comment by auraz** [comment]
Please also add steps for DevOps to setup JWT case in a step-by-step manner that can be copy-pasted or linked to them in future Jira task to connect LS with SF.

File: oauth2_client/models.py [hunk 10]
```diff
@@ -1,57 +1,192 @@
...
+
+        :return: dict
+        """
+        as_dict = model_to_dict(self, fields=["token", "token_type", "expires", "scope"])
+        as_dict["access_token"] = as_dict.pop('token')
+        if 'expires' in as_dict and as_dict['expires']:
+            # expiry info can be used by oauth2_session (client's 3rd party parent class)
+            expires_dt = as_dict.pop('expires')
+            as_dict["expires_at"] = int(datetime_to_float(expires_dt))
+            as_dict["expires_in"] = (expires_dt - timezone.now()).total_seconds()
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/models.py]: Given their current implementation, `expires_in` is enough. I will remove `expires_at` then from the code above. You are right, let's be lean. 
https://github.com/oauthlib/oauthlib/blob/master/oauthlib/oauth2/rfc6749/clients/base.py#L499-L504

File: oauth2_client/tests/test_flow_jwt.py (hunk 4 above)
Comment (fredmajor) [reply oauth2_client/tests/test_flow_jwt.py]: Installing pylint plugin for Pycharm so we don't have to waste not time on those anymore :)

File: oauth2_client/client.py (hunk 6 above)
Comment (auraz) [reply oauth2_client/client.py]: Discussed in slack, it looks like we do not need for callback to be optional.

File: oauth2_client/client.py (hunk 8 above)
Comment (fredmajor) [reply oauth2_client/client.py]: This is widely used as a separate function, for mocking in tests and now as a protection point with a circuit breaker.

File: oauth2_client/client.py [hunk 11]
```diff
@@ -1,37 +1,95 @@
...
+    OAuth2 client to make authorized HTTP(S) requests with OAuth2 token.
 
-    Wrapper around OAuth2Session to cache saervice host on the class instatiation and use it to transform
+    Wrapper around OAuth2Session to cache service host on the class instantiation and use it to transform
     relative urls to absolute ones.
 
+    Expired OAuth2 token can be refreshed by repeating the authorization flow. Token refresh
+    functionality is protected with circuit breakers, to avoid spamming authorization provider
+    and resource owner in case of system failure. Token expiry detection uses best-effort approach,
+    as there is no standardized way of informing the client of token expiry. The mechanism tries
```
Comment (auraz) [inline oauth2_client/client.py]: Because token expiration data is not always set?

File: oauth2_client/client.py [hunk 12]
```diff
@@ -1,37 +1,95 @@
...
-    OAuth2 client to make HTTP calls with OAuth2 token.
+    OAuth2 client to make authorized HTTP(S) requests with OAuth2 token.
 
-    Wrapper around OAuth2Session to cache saervice host on the class instatiation and use it to transform
+    Wrapper around OAuth2Session to cache service host on the class instantiation and use it to transform
     relative urls to absolute ones.
 
+    Expired OAuth2 token can be refreshed by repeating the authorization flow. Token refresh
+    functionality is protected with circuit breakers, to avoid spamming authorization provider
+    and resource owner in case of system failure. Token expiry detection uses best-effort approach,
```
Comment (auraz) [inline oauth2_client/client.py]: That is little unclear to me ` Token expiry detection uses best-effort approach`. We do not do any special detection  as I assume.

File: oauth2_client/client.py [hunk 13]
```diff
@@ -1,37 +1,95 @@
...
+    as there is no standardized way of informing the client of token expiry. The mechanism tries
+    to fetch new token and retries the request once. If unsuccessful - CircuitBreakerError is raised.
```
Comment (auraz) [inline oauth2_client/client.py]: fail_max=2 is it the same?

File: oauth2_client/tests/test_client.py [hunk 14]
```diff
@@ -1,84 +1,204 @@
 """
 Client tests.
+
+Local imports of model-related files are required for in-IDE test runs.
+Otherwise a model import is attempted before django.setup() call and
+an exception is thrown.
```
[the remark this answers is not in the document]
Comment (labstersteve) [reply oauth2_client/tests/test_client.py]: In future, `black` will decide ...  :-)

File: oauth2_client/tests/test_client.py [hunk 15]
```diff
@@ -1,84 +1,204 @@
...
-        Ensure no token is created if there is no expired one.
+        Ensure no token is fetched if there is a valid one.
         """
         from oauth2_client.models import AccessToken
-        from oauth2_client.client import get_client, TIMEOUT_SECONDS
-        from .factories import AccessTokenFactory, ApplicationFactory
+        from oauth2_client.client import get_client
+        from .factories import AccessTokenFactory, ApplicationFactory, fake_token
 
+        now = timezone.now()
```
Comment (labstersteve) [inline oauth2_client/tests/test_client.py]: Is this auxiliary variable needed? As far as I can see it's used only once
.

File: oauth2_client/tests/test_client.py [hunk 16]
```diff
@@ -1,84 +1,204 @@
...
         app = ApplicationFactory()
```
Comment (labstersteve) [inline oauth2_client/tests/test_client.py]: Similarly this one. Why not just use the vaklues inline in the call?

File: oauth2_client/client.py (hunk 11 above)
Comment (fredmajor) [reply oauth2_client/client.py]: resolved on a call

File: oauth2_client/client.py (hunk 12 above)
Comment (fredmajor) [reply oauth2_client/client.py]: resolved in a call

File: oauth2_client/client.py (hunk 13 above)
Comment (fredmajor) [reply oauth2_client/client.py]: resolved in a call

File: oauth2_client/client.py [hunk 17]
```diff
@@ -29,9 +38,58 @@ class OAuth2Client(OAuth2Session):
         > r = client.get('/api/license/1/detail/')
     """
 
-    def __init__(self, app, token):
-        self.service_host = app.service_host  # used to transform relative URLs to absolute
-        super(OAuth2Client, self).__init__(client_id=app.client_id, token=token)
+    """
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/client.py]: outdated due to introduction of 3rd party implementation of a circuit breaker pattern

File: oauth2_client/client.py (hunk 2 above)
Comment (fredmajor) [reply oauth2_client/client.py]: outdated due to introduction of 3rd party implementation of a circuit breaker pattern

File: oauth2_client/client.py (hunk 3 above)
Comment (fredmajor) [reply oauth2_client/client.py]: outdated due to introduction of 3rd party implementation of a circuit breaker pattern

File: oauth2_client/client.py (hunk 1 above)
Comment (fredmajor) [reply oauth2_client/client.py]: outdated

File: oauth2_client/fetcher.py [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+            raise
+
+    def requested_scope(self):
+        """
+        Get the scope to be requested from the provider in the auth flow.
+        """
+        scope = self.app.scope
+        return scope.split() if scope else []
+
+    def extract_received_scope(self, token, default=""):
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: for me:
`get_x` - the x is ready on the shelf and I just have to pick it up 
`extract_x` - I have to do some work to get my `x`
`x` - generic "give me x", telling me nothing on how much work is required

Removing the prefix per the review.

File: oauth2_client/fetcher.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+            token: token object or dict returned from auth provider
+            default: value to return when no scope info found
+
+        Returns:
+            scope string or default
+        """
+        received = getattr(token, "scope", None)
+        if received is None:
+            received = token.get('scope', default)
+
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: thanks, changing.

File: oauth2_client/fetcher.py [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+        # check if we got what we asked for
+        requested = self.requested_scope()
+        requested_scope = set(requested)
+        received_scope = set(received.split())
+        if requested_scope != received_scope:
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: refactored per Steve's comments.

File: oauth2_client/client.py [hunk 21]
```diff
@@ -43,63 +101,68 @@ def request(self, method, url, *args, **kwargs):  # pylint: disable=arguments-di
 
         Returns:
             Response object.
+
+        Raises:
+            TokenExpiredError: 1) when token expiry detected and `expired_callback` not specified;
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/client.py]: TokenExpiredError is defined by 3rd party code and used to signalize token expiry based on token's `expiry` timestamp, when available
I piggybacked to that approach and used the same exception to convey the same meaning - token expired.

File: oauth2_client/fetcher.py [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+            token_url=self.app.token_uri,
+            client_id=self.app.client_id,
+            client_secret=self.app.client_secret,
+            scope=self.requested_scope()
+        )
+
+
+def expires_or_none(raw_token):
+    """
+    Determine token's expiry date if available. The RFC isn't strict about this,
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: I like it, changing.

File: oauth2_client/fetcher.py [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+    if 'expires_in' in raw_token:
+        expires = timezone.now() + timedelta(seconds=raw_token['expires_in'])
+    elif 'expires_at' in raw_token:
+        expires = float_to_datetime(raw_token['expires_at'], tzinfo=pytz.UTC)
+    else:
+        expires = None
+
+    if expires and expires <= timezone.now():
+        log.warning(
+            "Received token already expired. Returning None for `expires` value. "
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: actually I think quite opposite would be nice to do here:
- if the token contains expiration timestamp (this is a token from oauth2_provider), raise a strong exception when the expiration timestamp is already in the past. This means 100% some problem with the provider.

[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: adding that exception

File: oauth2_client/client.py [hunk 24]
```diff
@@ -43,63 +101,68 @@ def request(self, method, url, *args, **kwargs):  # pylint: disable=arguments-di
...
+    Otherwise - new token is fetched from the auth provider by HTTP(S) and stored in the database.
+    The new token is then used for communication. Tokens are automatically refreshed by repeating
+    the authorization flow.
+
+    The client may occasionally return HTTP 401 when token expires during request processing on
+    Resource Owner side.
 
     Arguments:
-        service_name (str): name of the service make requests to f.e. licence.
+        app_name (str): name of the OAuth client application to make requests to f.e. license.
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/client.py]: I saw that in existing code and incorrectly assumed that's proper English. Fixing, good catch

File: oauth2_client/fetcher.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: restoring original version, as unit tests reminded it was what I actually need.
```
        received = getattr(token, "scope", None)
        if received is None:
            received = token.get('scope', default)
```
First of all, the token can be either an object with `.scope` property or a dict with `scope` key (provider specific thing)
I case of an object, there is a chance the .scope property is present, but equal None. Then I need to assign my default, an empty string (per django data conventions, as I store that in db later).  My solution accounts for all cases and works fine, that's not a code smell.

File: oauth2_client/fetcher.py [hunk 25]
```diff
@@ -0,0 +1,306 @@
+"""
+Components used to fetch (obtain) an OAuth2 token from authorization provider.
+Fetcher owns authorization flow logic. An `Application` object is bound to a
+fetcher and used as a source of credentials.
+
+Supported grant types:
+- Client Credentials (e.g. service-to-service communication)
+- JWT Bearer token (e.g. Salesforce)
+
+NOTE: by `Application` or `App` we mean an OAuth application and its
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: The only django-related usage of the name "Application" in this project is in the Readme, and goes something like: "It is a Django application without a full Django project"

All other usages assume an OAuth application and are consistent. I am hoping potential readers of those files will be OAuth-savvy, and "Application" is a well established term in the OAuth world: https://django-oauth-toolkit.readthedocs.io/en/latest/glossary.html#term-application 

To summarize, I see your point but I don't think introducing that extra verbosity would be a real help for the reader. You want to work with OAuth in a competent way? You are going to have to embrace that glossary I attached anyway..

File: oauth2_client/fetcher.py [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+
+    def fetch_raw_token(self):
+        """
+        Fetch token using JWT Bearer flow.
+
+        Returns:
+            dict: raw token from provider
+
+        Raises:
+            ValidationError: if the Application object we are fetching token
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: You are correct having seen other exceptions but that must have been in different places in the code. Extending the list here though, to account for all possibilities in this code. And yes, the circuit breaker signalizes its opening by raising `pybreaker.CircuitBreakerError`
```
        Raises:
            ValidationError: if the Application object we are fetching token
                for doesn't provide all required input data
            RequestException: from `requests` library
            JSONDecodeError: unparseable data received
```

File: oauth2_client/models.py [hunk 27]
```diff
@@ -1,57 +1,192 @@
...
+    raw_token = JSONField(
+        default=dict,
+        blank=True,
+        help_text="Token JSON object returned from provider. This is for debugging purposes and can"
+                  " be removed in the future."
+    )
+    token_type = models.TextField(
+        help_text="Token type, most likely Bearer"
+    )
+    application = models.ForeignKey(
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/models.py]: Refactoring lines to make them longer. Although, reading the autogenerated migration file, I see I am not the only one around with a fond for short lines ;)

File: oauth2_client/models.py [hunk 28]
```diff
@@ -1,57 +1,192 @@
...
+        """
+        The token is expired when 1) expiration info available AND 2) expiration datetime is in the past.
+        TIMEOUT_SECONDS margin is used to prevent token expiration issues during long-running
+        requests.
+        Apart from `expired`, the token could be `valid` or `unknown` (no expiration info available)
+
+        Returns:
+            bool: is token expired
+        """
+        # noinspection PyTypeChecker
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/models.py]: to silence a false alarm from Pycharm's linter. Removing.

File: oauth2_client/tests/resources/test.cert [hunk 29]
```diff
@@ -0,0 +1,15 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/tests/resources/test.cert]: adding info:
# needed for oauth2_client/tests/test_flow_jwt.py and oauth2_client/tests/test_crypto.py

File: oauth2_client/tests/test_client.py [hunk 30]
```diff
@@ -1,84 +1,204 @@
...
-    def test_no_token(self, get_token_mock):
+    @patch('oauth2_client.client.fetch_token')
+    def test_no_token(self, fetch_token_mock):
         """
         Ensure new token is created if there is no token.
         """
-        from oauth2_client.models import AccessToken
         from oauth2_client.client import get_client
-        from .factories import ApplicationFactory
+        from oauth2_client.models import AccessToken
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/tests/test_client.py]: good point sir, changing

[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/tests/test_client.py]: after talking to Steve - proceeding with a slightly different approach. Moving all problematic imports to one file, `ide_test_compat` and importing from there. 
This issue in general is a manifestation of a broader django-testing related tension, that I want to solve properly one day, once and for all. Meanwhile - I think approach with one file is good enough.

File: oauth2_client/tests/test_client.py (hunk 15 above)
Comment (fredmajor) [reply oauth2_client/tests/test_client.py]: this `now` is used in 2 places. Most of my `now`s are there for a reason.

File: oauth2_client/tests/test_client.py (hunk 16 above)
Comment (fredmajor) [reply oauth2_client/tests/test_client.py]: `app` was a justified remark indeed. Fixed across all tests.

File: oauth2_client/tests/test_flow_client_credentials.py (hunk 5 above)
Comment (fredmajor) [reply oauth2_client/tests/test_flow_client_credentials.py]: garbage from old tests, removed.

File: oauth2_client/utils/date_time.py [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    Convert a datetime object to a floating point timestamp.
+    Return a number of seconds elapsed from the UTC epoch.
+    Compatible with both Python 2 and 3.
+
+    If the input object is timezone-aware, the result includes timezone
+    difference between UTC and the timezone.
+
+    If the input object is timezone-naive, it's treated as UTC.
+    NOTE: This behaviour is different from the standard's library
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/utils/date_time.py]: Your sentence is an oxymoron. If a datetime object is a local timezone, it is timezone-aware by definition, so it contains tzinfo information. 
A datetime object without tzinfo is called a "timezone-naive"

File: oauth2_client/client.py (hunk 7 above)
Comment (fredmajor) [reply oauth2_client/client.py]: resolved in a call

File: oauth2_client/fetcher.py [hunk 32]
```diff
@@ -0,0 +1,306 @@
...
+        oauth = OAuth2Session(client=client)
+        return oauth.fetch_token(
+            token_url=self.app.token_uri,
+            client_id=self.app.client_id,
+            client_secret=self.app.client_secret,
+            scope=self.requested_scope()
+        )
+
+
+def expires_or_none(raw_token):
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: very unlikely this can be useful for anything else but the fetcher...

File: oauth2_client/fetcher.py [hunk 33]
```diff
@@ -0,0 +1,306 @@
...
+    if 'expires_in' in raw_token:
+        expires = timezone.now() + timedelta(seconds=raw_token['expires_in'])
+    elif 'expires_at' in raw_token:
+        expires = float_to_datetime(raw_token['expires_at'], tzinfo=pytz.UTC)
+    else:
+        expires = None
+
+    if expires and expires <= timezone.now():
+        log.warning(
+            "Received token already expired. Returning None for `expires` value. "
```
[the remark this answers is not in the document]
Comment (fredmajor) [reply oauth2_client/fetcher.py]: Good point. But replaced that with an exception, after thinking about that again.

File: oauth2_client/models.py [hunk 34]
```diff
@@ -136,7 +135,6 @@ class AccessToken(models.Model):
         null=True
     )
 
-    # Scope is a series of space delimited strings, eg. "read write"
```
Comment (labstersteve) [inline oauth2_client/models.py]: This information might be useful to casual readers of _fetcher.py_ unless the representation is hidden under some abstraction.

File: oauth2_client/tests/test_logging_cmd.py [hunk 35]
```diff
@@ -41,8 +41,8 @@ def setUpClass(cls):
     @classmethod
     def tearDownClass(cls):
         super(TestLoggingCommand, cls).tearDownClass()
-        # Restore test logging to ERROR only. We want clean output.
-        logging.disable(logging.WARNING)
+        # Restore test logging to ERROR and WARNING only. We want clean output.
```
Comment (labstersteve) [inline oauth2_client/tests/test_logging_cmd.py]: I very much approve of this. We need clarity on what constitutes a "failure" and to classify our messages accordingly.
